src/lib/NonLinearDutchDecayLib.sol

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.0;
+
+import {OutputToken, InputToken} from "../base/ReactorStructs.sol";
+import {NonLinearDutchOutput, NonLinearDutchInput, NonLinearDecay} from "../lib/NonLinearDutchOrderLib.sol";
+import {Util} from "./Util.sol";
+import {FixedPointMathLib} from "solmate/src/utils/FixedPointMathLib.sol";
+
+/// @notice helpers for handling dutch order objects
+library NonLinearDutchDecayLib {
+    using FixedPointMathLib for uint256;
+
+    /// @notice locates the current position on the curve and calculates the decay
+    function decay(NonLinearDecay memory curve, uint256 startAmount, uint256 decayStartBlock)
+        internal
+        view
+        returns (uint256 decayedAmount)
+    {
+        // handle current block before decay or no decay
+        if (decayStartBlock >= block.number) {
+            return startAmount;
+        }
+        uint256 blockDelta = block.number - decayStartBlock;
+        // iterate through the points and locate the current segment
+        for (uint256 i = 0; i < curve.relativeAmount.length; i++) {
```

Comment:
also because the max number of points on the curve is known and strictly increasing, we can optimize this via a binary search once we know if `blockDelta` is greater than or equal to the mid value in the curve

Reply:
I've implemented this in this [commit](https://github.com/Uniswap/UniswapX/pull/269/commits/59153044b910bb6d920f25cfc25d2469648421ad) however, I think it actually costs more gas for the length of points we're dealing with.
## Linear Search
![image](https://github.com/user-attachments/assets/6215996f-8266-4957-bf03-5951758cf4b5)

## Binary Search
<img width="916" alt="image" src="https://github.com/user-attachments/assets/5e6f5567-b6d1-4e45-89ed-436829e88ec9">

Reply:
Performed more extensive gas testing comparing linear with binary searching. Binary search only starts showing real benefits when we have 14 or more curve points. Opting for linear search since we will likely have less than 5 curve points defined and it's less bug prone.

<img width="740" alt="image" src="https://github.com/user-attachments/assets/7db348ff-51d1-4994-9d23-5b5aa7e77ba9">